forem/forem
# Detect animated images in articles asynchronously

## What type of PR is this? (check all applicable)


## Description

This spike aims to explore if detecting images server side is feasible.

The PR contains a `Articles::DetectAnimatedImages` service and its companion worker, which is triggered in a "after commit" callback for the `Article` model. What the worker does is the following:

- parse the `article.processed_html` field
- isolate all images that are not part of liquid tags (all the liquid tags we have cannot have animated images in them, even Twitter which renders animated images as videos with a preview frame)
- uses [fastimage](https://github.com/sdsykes/fastimage) to detect which ones are animated and adds a `data-animated=true` attribute to the HTML `<img>` tag
- busts the article cache to make sure the next person landing on the page will have the correct HTML

This should enable the frontend team to act accordingly when the article's content contains one or more animated images.

Some data:

- the average in the top 1000 articles on DEV is 6/7 images (probably less because I forgot to exclude the following tags in the SQL query: podcast, user and organization)
- the article that has the most has 200 images
- the `Articles::DetectAnimatedImagesWorker` takes around 2/3 seconds to run on an article with 7 images
- the same worker takes around 95 seconds to run on an article with 200 images

So in the worst case only people viewing an article in the 2 minutes following the publication will be served HTML without the `data-animated` attribute in the images.

## Related Tickets & Documents

Closes https://github.com/forem/forem/issues/13565

## QA Instructions, Screenshots, Recordings

1. Create or update articles adding animated images and static images (uploaded or linked, doesn't make a difference)
2. Refresh after a few seconds, check the `<img>` tags of the animated images have the `data-animated` attribute, make sure the other static images's tags are unchanged


https://user-images.githubusercontent.com/146201/118251472-eafe3b00-b4a7-11eb-91f4-567b94606230.mov

### UI accessibility concerns?

On average, in the first few seconds, users browsers cannot take advantage of the `data-animated` attribute

## Added tests?

- [x] No, and this is why: it's a spike, if we deemed this an approach we can follow, i'll add tests
